### DISCUSSION TITLE
How to close/terminate a client session in Wireguard mode

### DISCUSSION THREAD

**Discussion by phucnb** [mitmproxy/mitmproxy#6931] [Q&A]
I have set up a docker container in Wireguard mode. Users connect to the server via Wireguard app on their phone. Server will change the response for a special request. What i want is terminate/close the session after modify the response to avoid the situation users forget to turn off Wireguard VPN and let the VPN run all the time.

**Comment by emanuele-em** [comment]
Generally speaking I don't think this is a good thing to do but, if you want to kill only a single client session in wireguard you can use this command:
`wg set <interface> peer <key> remove`

You could use the [addons](https://docs.mitmproxy.org/stable/addons-examples/) power with something like this:
```python
def response(flow: http.HTTPFlow):
    assert flow.response
    if flow.request.path == "/something":
        flow.response.headers["something"] = "new-header"
    user_ip = flow.client_conn.peername[0]
    os.system(f"sudo /path/to/terminate_session.sh {user_ip}")
```

and terminate_session.sh could be:
```bash
USER_IP=$1
PEER=$(wg show wg0 allowed-ips | grep $USER_IP | awk '{print $1}')
if [ ! -z "$PEER" ]; then
    sudo wg set wg0 peer $PEER remove
fi
```

I didn't test these scripts and of course this is not a fully functional example, but I think it could be a good starting point

**Reply by phucnb** [reply]
> Generally speaking I don't think this is a good thing to do but, if you want to kill only a single client session in wireguard you can use this command: `wg set <interface> peer <key> remove`
> 
> You could use the [addons](https://docs.mitmproxy.org/stable/addons-examples/) power with something like this:
> 
> ```python
> def response(flow: http.HTTPFlow):
>     assert flow.response
>     if flow.request.path == "/something":
>         flow.response.headers["something"] = "new-header"
>     user_ip = flow.client_conn.peername[0]
>     os.system(f"sudo /path/to/terminate_session.sh {user_ip}")
> ```
> 
> and terminate_session.sh could be:
> 
> ```shell
> USER_IP=$1
> PEER=$(wg show wg0 allowed-ips | grep $USER_IP | awk '{print $1}')
> if [ ! -z "$PEER" ]; then
>     sudo wg set wg0 peer $PEER remove
> fi
> ```
> 
> I didn't test these scripts and of course this is not a fully functional example, but I think it could be a good starting point

Can you educate me why it's not a good thing to do?
I run the mitmproxy container in wireguard mode, I tried connect to the container and run some wg commands such as : wg show...but no response, i guess mitmproxy control the wireguard internally so I can't access and control the WG directly. Correct me if I am wrong please.
Thank you.

**Reply by mhils** [reply]
@emanuele-em: This suggestion unfortunately won't work on the mitmproxy end, because there is no `wg` daemon running in mitmproxy's WireGuard mode. :)

**Comment by mhils** [comment]
My proposal would be to just use https://docs.mitmproxy.org/stable/overview-features/#blocklist.

Or you write a custom addon where you record which users have visited your special URL, and then block all requests afterwards.

**Reply by phucnb** [reply]
I can block all the requests after they visited the special URL but it not solve the problem, all the requests still go to the proxy and got blocked. I want to terminate their session completely which they have to turn off and on vpn again to get a new session.

**Comment by mitmenjoyer** [comment]
@phucnb @mhils Was a solution ever found, I'm currently also struggling with this problem. I want to gracefully disconnect the user from the WireGuard tunnel once a request has been modified as well but can't figure out any good solution w/o making the client hang.
